<a href="https://colab.research.google.com/github/DiogoCanastro/Canastro/blob/main/Products_by_Location.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Extract `Location ID` and `Product ID` from the 'Location Product' sheet (sheet 3) of the '/content/Location Product MDs IBP 1.xlsx' file, then create a function that takes a `Location ID` as input and returns a list of associated `Product ID`s. Demonstrate this function with an example.

## Load Data from Excel Sheet

### Subtask:
Load the data from the 'Location Product' sheet (sheet number 3) of the '/content/Location Product MDs IBP 1.xlsx' file into a pandas DataFrame.


**Reasoning**:
To load the data from the specified Excel sheet into a pandas DataFrame and display its head and columns, I will first import the pandas library, then use `pd.read_excel` to load the data, and finally `df.head()` and `df.columns` to verify the load.



In [ ]:
import pandas as pd

# Load the data from the 'Location Product' sheet (sheet number 3)
df_location_product = pd.read_excel('/content/Location Product MDs IBP 1.xlsx', sheet_name='Location Product')

# Display the first few rows of the DataFrame
print("First 5 rows of the DataFrame:")
print(df_location_product.head())

# Display the columns of the DataFrame
print("\nColumns in the DataFrame:")
print(df_location_product.columns)

First 5 rows of the DataFrame:
  Unnamed: 0         SCNID         LOCID        PRDID  \
0          #       Version  Location ID*  Product ID*   
1          1  Base Version          0100       BT0001   
2          2  Base Version          0100       BT0003   
3          3  Base Version          0100       BT0010   
4          4  Base Version          0100       BT0011   

                           DDAOSH              ZSPAVERAGEWEEKS  \
0  Additional Order Spike Horizon  Average weeks - dep. demand   
1                             NaN                            2   
2                             NaN                            2   
3                             NaN                            2   
4                             NaN                            2   

                   ZSPAVGPERIODS               ZFIXEDPLAN  \
0  Average weeks - indep. demand  Cleanse Adj. Plan (Y/N)   
1                              0                        N   
2                              0              

**Reasoning**:
The previous output shows that the first row of the DataFrame contains the true headers ('Location ID*', 'Product ID*'), and the actual data begins from the second row. Also, the column names are currently `LOCID` and `PRDID`. To correctly extract `Location ID` and `Product ID`, I need to reload the Excel sheet by skipping the first row, select only the relevant columns, drop any rows with missing values in these columns, and then rename them to 'Location ID' and 'Product ID' as per the task.



In [ ]:
import pandas as pd

# Load the data again, skipping the first row to use the second row as headers
df_location_product_clean = pd.read_excel(
    '/content/Location Product MDs IBP 1.xlsx',
    sheet_name='Location Product',
    header=1 # Set header to 1 (0-indexed) to use the second row as column names
)

# Select only the 'Location ID*' and 'Product ID*' columns
df_location_product_clean = df_location_product_clean[['Location ID*', 'Product ID*']]

# Drop rows where either 'Location ID*' or 'Product ID*' is missing
df_location_product_clean.dropna(subset=['Location ID*', 'Product ID*'], inplace=True)

# Rename the columns for clarity as per the task description
df_location_product_clean.rename(columns={'Location ID*': 'Location ID', 'Product ID*': 'Product ID'}, inplace=True)

# Display the first few rows of the cleaned DataFrame
print("First 5 rows of the cleaned DataFrame:")
print(df_location_product_clean.head())

# Display the shape of the cleaned DataFrame to see the number of rows and columns
print("\nShape of the cleaned DataFrame:")
print(df_location_product_clean.shape)

First 5 rows of the cleaned DataFrame:
  Location ID Product ID
0        0100     BT0001
1        0100     BT0003
2        0100     BT0010
3        0100     BT0011
4        0100   BT202127

Shape of the cleaned DataFrame:
(37005, 2)


**Reasoning**:
Now that the data containing 'Location ID' and 'Product ID' has been loaded and cleaned, the next step is to create a function that takes a `Location ID` as input and returns a list of associated `Product ID`s, as specified in the main task. This function will use the `df_location_product_clean` DataFrame to filter products by location.



In [ ]:
def get_products_by_location(location_id):
    """
    Takes a Location ID as input and returns a list of associated Product ID's.
    """
    # Filter the DataFrame for the given location_id
    products_at_location = df_location_product_clean[df_location_product_clean['Location ID'] == location_id]

    # Extract the 'Product ID' column and convert it to a list
    product_ids = products_at_location['Product ID'].tolist()

    return product_ids

# Demonstrate the function with an example
example_location_id = '0100' # Using an example Location ID found in the data
associated_products = get_products_by_location(example_location_id)

print(f"Products associated with Location ID '{example_location_id}':")
print(associated_products)

# Display the number of products found for the example location
print(f"\nNumber of products found for Location ID '{example_location_id}': {len(associated_products)}")

Products associated with Location ID '0100':
['BT0001', 'BT0003', 'BT0010', 'BT0011', 'BT202127', 'BT30749', 'BT30846', 'BT30892', 'BT30937', 'BT31102', 'BT31131', 'BT33518', 'BT38600', 'BT39357', 'BT39358', 'BT39768', 'BT39770', 'BT4778', 'LA9950', 'RB0073', 'RB0323', 'RB0375', 'RB1248', 'RB1253', 'RB1260', 'RB1282', 'RB1304', 'RB1309', 'RB1325', 'RB1351', 'RB1446', 'RB1447', 'RB1449', 'RB1452', 'RB1460', 'RB1478', 'RB1515', 'RB1573', 'RB1576', 'RB1577', 'RB1589', 'RB1616', 'RB1655', 'RB1664', 'RB1718', 'RB1740', 'RB1748', 'RB1752', 'RB1755', 'RB1789', 'RB1805', 'RB1880', 'RB1896', 'RB1900', 'RB1921', 'RB1922', 'RB1935', 'RB1997', 'RB200010', 'RB200062', 'RB200113', 'RB200118', 'RB200124', 'RB200129', 'RB200156', 'RB200170', 'RB200272', 'RB200278', 'RB200279', 'RB200285', 'RB200301', 'RB200328', 'RB200341', 'RB200361', 'RB200398', 'RB200420', 'RB200421', 'RB200526', 'RB200527', 'RB200533', 'RB200563', 'RB200617', 'RB200637', 'RB200638', 'RB200661', 'RB200800', 'RB2011', 'RB201172', 'R

In [ ]:
example_location_id = '2661' # Using an example Location ID found in the data
associated_products = get_products_by_location(example_location_id)

print(f"Products associated with Location ID '{example_location_id}':")
print(associated_products)

Products associated with Location ID '2661':
['DD-1718', 'DD-205250', 'DD-205915', 'DD-205916', 'DD-205917', 'DD-208976', 'DD-211936', 'DD-213082', 'DD-215354', 'DD-218598', 'DD-221027', 'DD-223014', 'DD-223342', 'DD-224825', 'DD-230366', 'DD-231487', 'DD-233665', 'DD-233666', 'DD-234435', 'DD-240513', 'DD-242607', 'DD-242811', 'DD-242816', 'DD-242899', 'DD-242986', 'DD-2609', 'DD-2610', 'DD-2746', 'DD-2895', 'DD-2896', 'DD-2897', 'DD-30195', 'DD-30247', 'DD-30688', 'DD-3133', 'DD-31444', 'DD-31445', 'DD-3177', 'DD-34650', 'DD-34659', 'DD-4816', 'DD-4817', 'DD-7000', 'DD-7005', 'DD-7006', 'DD-7008', 'DD-7009', 'DD-7010', 'DD-7011', 'DD-7012', 'DD-7013', 'DD-7014', 'RB233665']


In [ ]:
import pandas as pd

# Convert the list of associated products to a DataFrame
df_associated_products = pd.DataFrame(associated_products, columns=['Product ID'])

# Define the output file path
output_file_path = 'associated_products.xlsx'

# Export the DataFrame to an Excel file
df_associated_products.to_excel(output_file_path, index=False)

print(f"Successfully exported {len(associated_products)} product IDs to '{output_file_path}'")

Successfully exported 53 product IDs to 'associated_products.xlsx'


## Final Task

### Subtask:
Summarize the created interface for retrieving product IDs by location ID and explain how to use it for further queries.


## Summary:

### Q&A

**Summarize the created interface for retrieving product IDs by location ID and explain how to use it for further queries.**

The created interface is a Python function named `get_products_by_location`. It takes a `Location ID` (as a string) as its sole argument. When called, it filters the pre-processed product-location mapping data to find all entries corresponding to the given location ID and then returns a Python list containing all associated `Product ID`s.

To use this interface, you would simply call the function with the desired location ID. For example, `associated_products = get_products_by_location('YOUR_LOCATION_ID')` would store a list of product IDs in the `associated_products` variable. This list can then be used for various purposes, such as:
*   Filtering other datasets containing product information.
*   Performing aggregations or calculations based on products at a specific location.
*   Displaying a catalog of available products for a given location in a user interface.

### Data Analysis Key Findings

*   The raw data from the 'Location Product' sheet (sheet 3) in the Excel file required cleaning because the actual column headers ('Location ID\*' and 'Product ID\*') were located in the second row, not the first.
*   After loading the data with `header=1`, selecting the relevant columns, and dropping rows with missing values, the cleaned DataFrame contained 37,005 rows and 2 columns.
*   A Python function, `get_products_by_location`, was successfully implemented to retrieve product IDs for a given location ID.
*   Demonstrating the function with the example `Location ID` '0100' successfully returned a list containing 3655 associated `Product ID`s.

### Insights or Next Steps

*   The `get_products_by_location` function provides an efficient way to query product information based on location, which can be integrated into larger data processing pipelines or reporting tools.
*   For further analysis, the list of product IDs returned by the function could be joined with other datasets (e.g., product details, sales data) to gain insights into inventory, sales performance, or product distribution per location.
